mm2src/trading_api/src/one_inch_api/types.rs

```diff
@@ -0,0 +1,416 @@
+#![allow(clippy::result_large_err)]
+
+use super::client::{ApiClient, QueryParams, UrlBuilder};
+use super::errors::ApiClientError;
+use ethereum_types::Address;
+use mm2_err_handle::mm_error::MmResult;
+use serde::Deserialize;
+
+const ONE_INCH_MAX_SLIPPAGE: f32 = 50.0;
+const ONE_INCH_MAX_FEE_SHARE: f32 = 3.0;
+const ONE_INCH_MAX_GAS: u128 = 11500000;
+const ONE_INCH_MAX_PARTS: u32 = 100;
+const ONE_INCH_MAX_MAIN_ROUTE_PARTS: u32 = 50;
+const ONE_INCH_MAX_COMPLEXITY_LEVEL: u32 = 3;
+
+macro_rules! push_if_some {
+    ($arr: expr, $k: expr, $v: expr) => {
+        if let Some(v) = $v {
+            $arr.push(($k, v.to_string()))
+        }
+    };
+}
+
+/// API params builder for swap quote
+#[derive(Default)]
+pub struct ClassicSwapQuoteParams {
+    src: String,
+    dst: String,
+    amount: String,
+
+    // Optional fields
+    fee: Option<f32>,
+    protocols: Option<String>,
+    gas_price: Option<String>,
+    complexity_level: Option<u32>,
+    parts: Option<u32>,
+    main_route_parts: Option<u32>,
+    gas_limit: Option<u128>,
+
+    include_tokens_info: Option<bool>,
+    include_protocols: Option<bool>,
+    include_gas: Option<bool>,
+    connector_tokens: Option<String>,
+}
+
+impl ClassicSwapQuoteParams {
+    pub fn new(src: String, dst: String, amount: String) -> Self {
+        Self {
+            src,
+            dst,
+            amount,
+            ..Default::default()
+        }
+    }
+
+    pub fn with_fee(mut self, fee: Option<f32>) -> Self {
+        self.fee = fee;
+        self
+    }
+    pub fn with_protocols(mut self, protocols: Option<String>) -> Self {
+        self.protocols = protocols;
+        self
+    }
+    pub fn with_gas_price(mut self, gas_price: Option<String>) -> Self {
+        self.gas_price = gas_price;
+        self
+    }
+    pub fn with_complexity_level(mut self, complexity_level: Option<u32>) -> Self {
+        self.complexity_level = complexity_level;
+        self
+    }
+    pub fn with_parts(mut self, parts: Option<u32>) -> Self {
+        self.parts = parts;
+        self
+    }
+    pub fn with_main_route_parts(mut self, main_route_parts: Option<u32>) -> Self {
+        self.main_route_parts = main_route_parts;
+        self
+    }
+    pub fn with_gas_limit(mut self, gas_limit: Option<u128>) -> Self {
+        self.gas_limit = gas_limit;
+        self
+    }
+    pub fn with_include_tokens_info(mut self, include_tokens_info: Option<bool>) -> Self {
+        self.include_tokens_info = include_tokens_info;
+        self
+    }
+    pub fn with_include_protocols(mut self, include_protocols: Option<bool>) -> Self {
+        self.include_protocols = include_protocols;
+        self
+    }
+    pub fn with_include_gas(mut self, include_gas: Option<bool>) -> Self {
+        self.include_gas = include_gas;
+        self
+    }
+    pub fn with_connector_tokens(mut self, connector_tokens: Option<String>) -> Self {
+        self.connector_tokens = connector_tokens;
+        self
+    }
```

Comment:
I am not a fan of this pattern personally, but if we really want to do this I guess the best way would be creating macro which implements this pattern automatically without taking too much space in the module.

Reply:
> I am not a fan of this pattern personally, but if we really want to do this I guess the best way would be creating macro which implements this pattern automatically without taking too much space in the module.

really like this idea

Reply:
made a macro, great idea, ty

Reply:
https://github.com/KomodoPlatform/komodo-defi-framework/pull/2222#discussion_r1769642501